### PR TITLE
PatientFinders find patients in OpenMRS that match cases in CommCare

### PR DESCRIPTION
**Pull request by kaapstorm**
Getting there, @dannyroberts 

cc @sheelio buddy @mkangia 

:blowfish:

### CODE REVIEW

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [reply corehq/motech/openmrs/finders.py]: Yeah but no, Mr Stickler. That numerator is a float, so that's obviously [true division](https://www.python.org/dev/peps/pep-0238/).

Do you have any strong feelings about "future division" vs casting, reviewers?

Comment (dannyroberts) [reply corehq/motech/openmrs/finders.py]: I think it's just saying you should include

```
from __future__ import division
```
https://github.com/janjur/readable-pylint-messages/blob/master/README.md#old-division

as a way to ensure that this remains python3 compatible. Even though in this case what you wrote here I think would be python3-compatible, not casting and using the `__future__` import marks this file's division as python3-compatible so when we make the switch no one has to think about whether this particular line is or isn't compatible. Does that make sense? I think in general that's why we turned on stricter py3-compat linter errors.

File: corehq/motech/openmrs/finders.py
Comment (mkangia) [inline corehq/motech/openmrs/finders.py]: nit: is this return required because this a decorator and should return the `class_`?

File: corehq/motech/openmrs/finders.py
Comment (mkangia) [inline corehq/motech/openmrs/finders.py]: Should we use `get` here `case_config['person_properties']` instead to avoid failure in case the key is missing? Similarly for other 3 `property_map` iteration after this..

File: corehq/motech/openmrs/finders.py
Comment (mkangia) [inline corehq/motech/openmrs/finders.py]: `/ws/rest/v1/patient` looks like a constant? Can just extract it out.

File: corehq/motech/openmrs/repeater_helpers.py
Comment (mkangia) [inline corehq/motech/openmrs/repeater_helpers.py]: Can this return a message/exception that the `finder` was not found? Combining this with `None` when no patients are found would not make it clear why there were no records found.

**Review by mkangia — APPROVED** [review]
Overall looks good.. Didn't understand the use case and overall purpose so reviewed till the extent i could.

**Comment by sheelio** [comment]
@kaapstorm 
Still trying to figure out exactly how this is supposed to work.  My understanding was:
- If there is a match on case.openmrs_id, this is 100% trustworthy and we should just use it
- If there is a match on something like bahmni_id or household_id, we should do a secondary validation on something like DOB, first_name+last_name
- Once a match happens, we update the case with the OpenMRS ID (so in the future we can just rely on it)

Is this what the code is doing?

**Comment by kaapstorm** [comment]
@sheelio, It does the first two steps, but not yet the last one. 

The first step, "If there is a match on case.openmrs_id", happens in the [`get_patient` function](https://github.com/dimagi/commcare-hq/pull/18904/commits/600d41ab303e41c60a41905d9ac973f594c9d980#diff-ad1809f9bda087777a6c93629dda7a67R248). It uses the ID matchers that are part of the normal OpenMRS case config. If "openmrs_id" (or any other 100% certain ID matcher that might have been configured) does match, then we use that patient. 

Otherwise we go to Step 2, which tries to find the patient. [Properties are compared in order](https://github.com/dimagi/commcare-hq/pull/18904/commits/8fb3bb9a55982f00bb06aa9bd05ac1baf487d558#diff-b2bb772d97ec070579013d8ebebf68feR16) and we validate all of them. First, `bahmni_id`, then `household_id`, then `dob`, `first_name`, `last_name` and finally `municipality`. 

It's a little more complicated than what we discussed because we can't search on all properties, only IDs and last name, and searches are expensive but comparisons are cheap. So we search against each of the searchable properties, and compare all the weighted properties. Found patients are stored in a dictionary keyed on their OpenMRS ID to prevent duplicate search results.

Instead of just accepting all patients that score higher than 1 (the default but configurable value of `threshold`), we have a final comparison. Imagine there are 2 John Smith's from Somerville who are born on Jan 1 1980. They will both score 1 for DOB + full name + municipality. But one of them has a matching household_id and the other doesn't. So his score is 1.9 and the other is 1.0. If one patient's score is more that 10% (by default but configurable) better than another, we pick him. If there is less than a 10% difference we return both (and neither will be selected, but they will be logged).

We will log all these results, so we can see whether our weights and thresholds make sense. 

Let me know if that answers your question, and if you have any concerns with that approach.

If we can get a real or typical sample dataset, and ignore the OpenMRS IDs, and run these comparisons, and then compare our results with known matches, we could refine (programmatically?) our weights and 10% margin, until we get as close as possible to the known matches, while minimising type 2 / false positive errors.

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [inline corehq/motech/openmrs/finders.py]: Example value for `patient['identifiers']`:
```json
  "identifiers": [
    {
      "display": "Patient Identifier = GAN200036",
      "uuid": "c0d20f06-e3c9-406e-95dd-ca55cc2e3818",
      "identifier": "GAN200036",
      "identifierType": {
        "uuid": "81433852-3f10-11e4-adec-0800271c1b75",
        "display": "Patient Identifier",
        "links": [ ,,, ]
      },
      "location": null,
      "preferred": true,
      "voided": false,
      "links": [ ,,, ],
      "resourceVersion": "1.8"
    }
  ],
```

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [inline corehq/motech/openmrs/finders.py]: `self.save_match_id` will save the patient's value of the identifier(s) listed in the `OpenmrsCaseConfig`'s "id_matchers" to the case, so that next time the case's patient will be matched immediately. An example value of OpenmrsCaseConfig.id_matchers would look like:
```json

  "id_matchers": [
    {
      "case_property": "openmrs_id", 
      "identifier_type_id": "8d793bee-c2cc-11de-8d13-0010c6dffd0f", 
      "doc_type": "IdMatcher"
    }
  ], 
```

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [reply corehq/motech/openmrs/finders.py]: `case_config` is an instance of [OpenmrsCaseConfig](https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/openmrs_config.py#L90) so it will always have `person_properties` ... but you make a good point, because that is not obvious. Maybe it would be better to use it as an attribute, `case_config.person_properties`, instead of a key.

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [reply corehq/motech/openmrs/finders.py]: Hmmm. That's true, [but][1] [there][2] [are][3] [a][4] [lot][5] [of][6] [endpoints][7] [like][8] [it][9] ... [and][10] [we][11] [only][12] [refer][13] to each of them once ... except that one, now that I'm going through them! I'll reuse [search_patients()][14], so at least all the endpoints are in the same module.


  [1]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L89
  [2]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L98
  [3]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L129
  [4]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L141
  [5]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L145
  [6]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L160
  [7]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L167
  [8]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L184
  [9]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L200
  [10]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L213
  [11]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L223
  [12]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L359
  [13]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L363
  [14]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L166

File: corehq/motech/openmrs/finders.py
Comment (kaapstorm) [reply corehq/motech/openmrs/finders.py]: Yes, exactly.

```python
@foo
class Bar:
    pass
```
is [the same](https://www.python.org/dev/peps/pep-3129/#id11) as
```python
class Bar:
    pass
Bar = foo(Bar)
```

**Comment by mkangia** [comment]
@kaapstorm is this waiting on me for review?

**Comment by kaapstorm** [comment]
@mkangia no. I just want a bit more testing, which keeps getting pushed back by other things. -- Hopefully tomorrow. 🤞

**Comment by mkangia** [comment]
cool. @kaapstorm please let me know when and if you want me to take another look

**Comment by kaapstorm** [comment]
@mkangia I'll rebase off #19525 ... and keep testing. (Nice to finally be making progress on this.)

File: corehq/motech/openmrs/jsonpath.py
Comment (kaapstorm) [reply corehq/motech/openmrs/jsonpath.py]: Ignoring. Implemented `__eq__` to conform to other JSONPath subclasses, but none of them implement `__hash__`, so I'm not going to either.

**Comment by dannyroberts** [comment]
@kaapstorm looks like there was some recent work here, but not sure whether you're still adding things or you're just waiting for a final review. Just let me know if/when you need a review and I'll spring into action

**Comment by kaapstorm** [comment]
Hey @dannyroberts the conversation seems to have migrated to PR #19836 which builds on this. I think we're pretty close to merging, so if you have a chance, it would be great if you could eyeball it.

**Pull request merged**
